src/app/scripts/cac/control/cac-control-directions.js

```diff
@@ -98,7 +102,7 @@ CAC.Control.Directions = (function (_, $, moment, Control, Routing, UserPreferen
      * Set user preferences before planning trip.
      * Throttled to cut down on requests.
      */
-    var planTrip = _.throttle(function() {  // jshint ignore:line
+    var planTrip = _.debounce(function() {  // jshint ignore:line
```

Comment:
What are the pros and cons of this?  On the pro side, it will drop more useless requests in the case of fast switching back and forth (though that doesn't seem like a huge real-world case) or someone clicking into directions then immediately changing mode options.  But the downside is that for simple interactions where the user clicks into directions with the desired mode options already selected--i.e. there's only one request total--this will make it .75 seconds slower.

We could set the option `leading=true` to avoid that, though that would mean the "get directions, switch mode" cases would always produce 2 requests rather than sometimes dropping the first.  It would still save requests vs. the current throttle in a rapid-toggling situation.

Reply:
See notes section above: there is an issue with the lodash throttle implementation.

In testing, I haven't found any noticeable difference in responsiveness; the actual query time is much longer than the debounce window.

Reply:
I had to go back and read [the debounce vs. throttle explainer](https://css-tricks.com/debouncing-throttling-explained-examples/) again and by the time I had finished I had somehow dropped the bit about the throttle bug from my brain.  And yeah, I think debounce makes more sense here, since even without the bug in throttle, sending requests in the middle of a stream of rapid invocations wouldn't be useful.

But I still think debounce-with-leading-edge seems preferable.  It still wouldn't be the same as throttle (which is, for Lodash, debounce-with-leading-edge-and-max-wait), though it would be susceptible to a similar effect, where if the user waits 751ms between requests the debounced grouping would close and send the trailing request and 1ms later the leading request of the next batch would go.

Running a few simple directions lookups on the live site, I'm getting responses in the 400-500ms range.  For longer trips, I'm seeing 1-1.25 seconds.  Adding 750ms to all of those seems to me like a potentially significant/noticeable difference, and I don't think it's worth it to add that lag to what seems to me like a core/common interaction to save load in what seem like less common scenarios.

Reply:
Added leading edge.

Reply:
On adding leading edge, the issue is reintroduced:

![image](https://user-images.githubusercontent.com/960264/36798874-ce83ea20-1c79-11e8-915c-35f6ca059bb4.png)

Reply:
Since there's logic now to throw away outdated, pending requests, maybe simply reducing the debounce timeout would be fine.

Reply:
Even with calls made in quick succession, as can happen with `leading` turned on, it shouldn't be possible for stale results to get displayed because the calls will always be in order and superseded ones should be getting canceled.  I think I see how it's happening.  I set `leading` to true and added some logging, then tested with [this trip](http://localhost:8024/?origin=40.338648%2C-75.11031&originText=Boxwood%20Cir%20%26%20Cold%20Spring%20Creamery%20Rd%2C%20Doylestown%2C%20Pennsylvania%2C%2018902&mode=WALK&maxWalk=482802&wheelchair=false&bikeOptimize=GREENWAYS&arriveBy=false&dateTime=&destination=39.878307%2C-75.268882&destinationText=John%20Heinz%20National%20Wildlife%20Refuge&waypoints=40.2171598%2C-75.0112152%3B40.0486427%2C-75.0359344), which is significantly faster without transit than with (maybe that's true in most cases, but just to have a working example written down...).  Start from no-transit, I click three times quickly (on-off-on), wait the moment it takes for the debounce timeout to pass and the spinner to appear, then click once more to turn transit back off.  So it sends the first click (with transit) as the first leading request, ignores the second click, sends the third (no transit) as the trailing request, then sends the fourth click (with transit) as the leading request of the next debounce grouping.  That leaves me in a state where I should be getting walking directions but the transit directions are reliably coming back last and showing up.

The console looks like this:
![image](https://user-images.githubusercontent.com/6598836/36862004-98dffa28-1d53-11e8-9ca1-beba08102bec.png)

I've annotated it with my interpretation of which requests the lines are talking about.  The interesting part is that when the trailing request comes back, it should have been canceled by the 2nd leading request, and in fact `planTripRequest` is null, but the `.then` callback is still firing.  My interpretation is that when `planTripRequest` gets set to `null`, that doesn't destroy the object, and the callback is still attached to its success handler.  On this theory, the fact that the right itinerary is always winning currently has mainly to do with the 750ms minimum gap between requests being enough to keep things in order.

Adding a check and bailing out if `planTripRequest` is falsy at the top of the `then` function appears to cause the final selection to always end up being the one displayed.

Reply:
Modified to never nullify the trip request, since rejecting an already completed request does nothing and does not error, and it seems to resolve the leading edge issue.

Reply:
Yay!  Repeating the above with the new code, I now see two rejections--i.e. Leading 1 gets rejected by Trailing, which in turn gets rejected by Leading 2--which is as it should be and results in it showing the right thing.